### PR TITLE
Added support for Azure Bastion.

### PR DESCRIPTION
**Pull request by whites11**
/kind feature

**What this PR does / why we need it**:

Towards: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/165

This PR adds support for the first of 3 bastion host implementations for CAPZ: the one using [Azure Bastion](https://azure.microsoft.com/en-us/services/azure-bastion/).

The PR adds a new field to the `AzureCluster` CR that looks like this:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: AzureCluster
metadata:
  name: d3m3v
  namespace: default
spec:
  bastionSpec:
    azureBastion:
      name: "azurebastion1"
      subnet: *SubnetSpec
      publicIP:
        name: "azurebastion1-publicIP"
```

by default the `bastionSpec` field is empty and the `azureBastion` field is therefore null (that means the feature is disabled by default).
User can simply set a value of the `azureBastion` field to enable the feature (other fields are optional and defaulted).

Once the feature is enabled, removing it from a cluster is unsupported (see [slack thread](https://kubernetes.slack.com/archives/CEX9HENG7/p1617883958183000)).

It is possible to enable the azureBastion feature on an existing v1alpha4 cluster though.

**Special notes for your reviewer**:

Not sure where to put documentation about the feature, guidance would be appreciated.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

No image version changes.

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Added support for using [Azure Bastion](https://azure.microsoft.com/en-us/services/azure-bastion/) to get console access to virtual machines in the cluster through the Azure Portal.
```

### CODE REVIEW

**Comment by shysank** [comment]
@whites11 can you add the copyright info to any new files you created. I think the `pull-cluster-api-provider-azure-verify` test  is failing because of that.

**Comment by whites11** [comment]
> @whites11 can you add the copyright info to any new files you created? I think `pull-cluster-api-provider-azure-verify` is failing because of that.

Fixed, thx

**Comment by whites11** [comment]
> @whites11 can you add the copyright info to any new files you created? I think `pull-cluster-api-provider-azure-verify` is failing because of that.

Not sure how to fix the apidiff job. Any hints?

**Comment by shysank** [comment]
> Not sure how to fix the apidiff job. Any hints?

It doesn't have to be fixed. They are just warnings when there are changes in api.

File: api/v1alpha3/azurecluster_conversion.go
Comment (devigned) [inline api/v1alpha3/azurecluster_conversion.go:63]: Could this not be pull from `restored`?

File: api/v1alpha3/azurecluster_conversion.go
Comment (devigned) [inline api/v1alpha3/azurecluster_conversion.go:96]: Looks like there are 2 calls to utilconversion.MashalData. There should only be one.

File: api/v1alpha3/azurecluster_conversion.go
Comment (devigned) [inline api/v1alpha3/azurecluster_conversion.go:99]: I believe this should be marshaled as part of the `utilconversion.MarshalData` can and be available for restoration via `restored` above. I don't thing this block is needed.

**Review by devigned — CHANGES_REQUESTED** [review]
I think the conversion code can be slimmed up a bit. I don't think the extra bastion annotation is needed as the entire converted structure should be available via the `utilconversion` annotation.

File: api/v1alpha3/azurecluster_conversion.go
Comment (whites11) [reply api/v1alpha3/azurecluster_conversion.go]: Oh, thanks for opening my eyes. I struggled to make the conversion test to pass and it was so easy after all.
Should hopefully be ok now.

File: api/v1alpha3/azurecluster_conversion.go
Comment (devigned) [inline api/v1alpha3/azurecluster_conversion.go:60]: nit: please remove the space between line 59 and 61.

**Review by devigned — CHANGES_REQUESTED** [review]
One small nit, but overall the PR looks great!

I do have a concern about not having an e2e test to ensure the functionality doesn't regress. Would you consider adding a bastion to the existing e2e cluster deployment (perhaps, `./templates/test/ci/prow`) and a test verifying successful deployment of the bastion.

I wonder if the bastion shouldn't be part of the default template... @CecileRobertMichon wdyt?

File: api/v1alpha3/azurecluster_conversion.go
Comment (whites11) [reply api/v1alpha3/azurecluster_conversion.go]: Blank line removed.
Regarding the e2e test, I will work on that.

**Comment by CecileRobertMichon** [comment]
I promise to take a closer look tomorrow but in the meantime, would love to see some user-facing docs as part of this PR.

re: testing/flavors, I'm worried we're adding too many different flavors / test for each feature, we should start combining some. In what use case does it make the most sense for a user to turn this on? a private API server w/ no ssh maybe? or is it useful even when ssh access is possible? I wouldn't add it to the "default" flavor template if it's going to incur some extra cost for every user (especially quick start users), but it might make sense to add it to the default-prow template (or just the private cluster template if that makes more sense in terms of use case).

**Comment by whites11** [comment]
During office hours earlier today we agreed enabling the azure bastion feature in the "internal networking" e2e test and adding additional checks into that test rather than having a dedicated e2e test just for this feature.

**Comment by whites11** [comment]
Can anybody please help me a little bit with the e2e tests?
What do I need to test? I expected to test the fact that when the azure bastion feature is enabled in a cluster, the `Azure Bastion` cloud object gets created and becomes healthy,
Problem is that I don't see any example of this kind of tests in the code.
Can anybody please provide some guidance?

**Comment by devigned** [comment]
@whites11 I don't think we are checking the Azure resources are created, but rather the functionality is working. For example, if we apply a machine deployment, we would then check if we can run workloads on those nodes. Perhaps, the same strategy could be used in the bastion. One example that comes to mind is: "Can you ssh into the bastion and then tunnel to one of the private nodes?". wdyt?

**Comment by devigned** [comment]
Hmm... here's an example of setting up an Azure client and calling out to Azure Resource Manager (ARM) to get event logs. Perhaps, use a similar pattern where you create a client and request the bastion resource from ARM and verify the properties you expect are set. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e6edff7b16eab72bca70dc4d95b82feb407647ca/test/e2e/e2e_suite_test.go#L181-L232

I was thinking about a VM bastion scenario, not the Azure Bastion. My fault.

**Comment by whites11** [comment]
I don't get why `pull-cluster-api-provider-azure-verify` fails: running `make generate` locally doesn't change files

**Comment by devigned** [comment]
@whites11 perhaps, wipe out your `./hack/tools/bin` directory. Maybe it's using an older version of a bin than the CI server.

**Comment by whites11** [comment]
> @whites11 perhaps, wipe out your `./hack/tools/bin` directory. Maybe it's using an older version of a bin than the CI server.

Didn't work unluckily.
Can it be related to this error I get in the output (both locally and in the CI)?

```
E0504 12:44:05.503496    1728 conversion.go:193] Rename function sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3 Convert_v1alpha3_ManagedDisk_To_v1alpha4_ManagedDiskOptions -> Convert_v1alpha3_ManagedDisk_To_v1alpha4_ManagedDiskParameters to match expected conversion signature
E0504 12:44:05.503732    1728 conversion.go:193] Rename function sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3 Convert_v1alpha4_ManagedDiskOptions_To_v1alpha3_ManagedDisk -> Convert_v1alpha4_ManagedDiskParameters_To_v1alpha3_ManagedDisk to match expected conversion signature
```

**Comment by devigned** [comment]
> Can it be related to this error I get in the output (both locally and in the CI)?

I don't think so. It's only a warning. I'm seeing it locally too.

**Comment by devigned** [comment]
@whites11 I'm seeing the issue in one of my PRs too. I'm looking into the resolution.

**Comment by whites11** [comment]
> @whites11 I'm seeing the issue in one of my PRs too. I'm looking into the resolution.

OK, thanks for the heads up

**Comment by CecileRobertMichon** [comment]
Did you rebase on the latest?

**Comment by devigned** [comment]
Here's the diff I get in #1332 via [the verify build log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1332/pull-cluster-api-provider-azure-verify/1389627379143413760/build-log.txt)

```
diff --git a/api/v1alpha3/zz_generated.conversion.go b/api/v1alpha3/zz_generated.conversion.go
index f124d9a7..ab9dfbd7 100644
--- a/api/v1alpha3/zz_generated.conversion.go
+++ b/api/v1alpha3/zz_generated.conversion.go
@@ -1299,7 +1299,7 @@ func autoConvert_v1alpha4_NetworkSpec_To_v1alpha3_NetworkSpec(in *v1alpha4.Netwo
 func autoConvert_v1alpha3_OSDisk_To_v1alpha4_OSDisk(in *OSDisk, out *v1alpha4.OSDisk, s conversion.Scope) error {
 	out.OSType = in.OSType
 	out.DiskSizeGB = in.DiskSizeGB
-	// WARNING: in.ManagedDisk requires manual conversion: inconvertible types (./api/v1alpha3.ManagedDisk vs *sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4.ManagedDiskParameters)
+	// WARNING: in.ManagedDisk requires manual conversion: inconvertible types (sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3.ManagedDisk vs *sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4.ManagedDiskParameters)
 	out.DiffDiskSettings = (*v1alpha4.DiffDiskSettings)(unsafe.Pointer(in.DiffDiskSettings))
 	out.CachingType = in.CachingType
 	return nil
@@ -1308,7 +1308,7 @@ func autoConvert_v1alpha3_OSDisk_To_v1alpha4_OSDisk(in *OSDisk, out *v1alpha4.OS
 func autoConvert_v1alpha4_OSDisk_To_v1alpha3_OSDisk(in *v1alpha4.OSDisk, out *OSDisk, s conversion.Scope) error {
 	out.OSType = in.OSType
 	out.DiskSizeGB = in.DiskSizeGB
-	// WARNING: in.ManagedDisk requires manual conversion: inconvertible types (*sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4.ManagedDiskParameters vs ./api/v1alpha3.ManagedDisk)
+	// WARNING: in.ManagedDisk requires manual conversion: inconvertible types (*sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4.ManagedDiskParameters vs sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3.ManagedDisk)
 	out.DiffDiskSettings = (*DiffDiskSettings)(unsafe.Pointer(in.DiffDiskSettings))
 	out.CachingType = in.CachingType
 	return nil
```

**Comment by CecileRobertMichon** [comment]
Opened https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1363 to fix the conversion func name and get rid of the warning.

Are you rebased on latest @devigned ? I don't get any diff running make generate from master

**Comment by devigned** [comment]
I am rebased on master. I also added the same clean up as #1363, but am still seeing the diff only in CI.

**Comment by whites11** [comment]
> Did you rebase on the latest?

I did it a few hours ago, will do it again tomorrow

**Comment by devigned** [comment]
I was still having an issue generating conversions when outside of the GOPATH. When inside of the GOPATH, everything works fine. 

@whites11 are you working from outside of your GOPATH? If so, can you repro the same thing from inside your GOPATH?

**Comment by whites11** [comment]
> I was still having an issue generating conversions when outside of the GOPATH. When inside of the GOPATH, everything works fine.
> 
> @whites11 are you working from outside of your GOPATH? If so, can you repro the same thing from inside your GOPATH?

Witnessing the same. Running `make generate` from within GOPATH leads to some changes.
It also creates a weird folder tree within the repo:

```
$ git status
On branch azure-bastion
Your branch is up to date with 'origin/azure-bastion'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   templates/cluster-template-azure-bastion.yaml

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        github.com/

no changes added to commit (use "git add" and/or "git commit -a")
```

```
$ ls -R github.com/
github.com/:
kubernetes-sigs

github.com/kubernetes-sigs:
cluster-api-provider-azure

github.com/kubernetes-sigs/cluster-api-provider-azure:
api  exp

github.com/kubernetes-sigs/cluster-api-provider-azure/api:
v1alpha3

github.com/kubernetes-sigs/cluster-api-provider-azure/api/v1alpha3:
zz_generated.conversion.go

github.com/kubernetes-sigs/cluster-api-provider-azure/exp:
api

github.com/kubernetes-sigs/cluster-api-provider-azure/exp/api:
v1alpha3

github.com/kubernetes-sigs/cluster-api-provider-azure/exp/api/v1alpha3:
zz_generated.conversion.go
```

**Comment by whites11** [comment]
Alright, e2e are finally green I guess that should be it? Thoughts anyone?

**Review by devigned — APPROVED** [review]
/lgtm

Congrats on adding the e2e test. Well done!

File: azure/services/bastionhosts/azurebastion.go
Comment (shysank) [inline azure/services/bastionhosts/azurebastion.go:107]: I don't think we can return here since we'll need to check for other resources

File: azure/scope/cluster.go
Comment (shysank) [inline azure/scope/cluster.go:259]: IIUC this means azureBastionSubnet will be reconciled in `subnetsSvc` as well as in `bastionSvc`. Is this intentional?

**Review by shysank — COMMENTED** [review]
A couple of nits and clarifications. I'm fine with opening up a followup pr for the nits as this has been around for a while, and don't want to get lgtm removed. Just want to make sure about the subnet and reconcileDelete clarifications before approving.

File: config/default/manager_image_patch.yaml
Comment (CecileRobertMichon) [inline config/default/manager_image_patch.yaml:11]: revert this

File: config/default/manager_pull_policy.yaml
Comment (CecileRobertMichon) [inline config/default/manager_pull_policy.yaml:11]: this too

**Review by CecileRobertMichon — COMMENTED** [review]
/hold

File: api/v1alpha4/azurecluster_types.go
Comment (whites11) [reply api/v1alpha4/azurecluster_types.go:71]: yes makes sense

File: api/v1alpha4/azurecluster_default.go
Comment (whites11) [reply api/v1alpha4/azurecluster_default.go:218]: default is no security group and no role.
The fields are there because they belong to the `SubnetSpec` type.

File: azure/services/bastionhosts/azurebastion.go
Comment (whites11) [reply azure/services/bastionhosts/azurebastion.go:107]: Good catch, thanks

File: config/default/manager_image_patch.yaml
Comment (whites11) [reply config/default/manager_image_patch.yaml]: done

File: config/default/manager_pull_policy.yaml
Comment (whites11) [reply config/default/manager_pull_policy.yaml]: done

File: azure/scope/cluster.go
Comment (whites11) [reply azure/scope/cluster.go:268]: This is actually a very good point! Not sure how the PR ended up managing subnet and public IP that way, but I removed my handlers and let the other services reconcile the subnet and public IP.
It feels much better now, thanks @shysank

Comment (shysank) [reply azure/scope/cluster.go:268]: Can we leave bastion subnets and bastion public ip reconciliation to `subnteSvc` and `publicIpSvc` respectively and have `bastionSvc` only reconcile bastion host? Since we reconcile subnets and publicIps  are reconciled before bastion, you should not have to check for their existence, and it will make the code much simpler as well.  wdyt?

Comment (shysank) [reply azure/scope/cluster.go:268]: @whites11 Just checking to see if you need any help or is this something you feel is not needed? 
cc: @CecileRobertMichon for more 👀

Comment (nader-ziada) [reply azure/scope/cluster.go:268]: makes sense to reconcile bastion subnets and bastion public ip by the subnteSvc and publicIpSvc

**Comment by nader-ziada** [comment]
tested the flow manually and confirmed the e2e tests works as expected, I don't see any outstanding comments 

/lgtm
/approve

**Comment by nader-ziada** [comment]
@whites11 you are going to have to generate again to get the verify to pass after the update to capi PR

**Comment by nader-ziada** [comment]
> @whites11 you are going to have to generate again to get the verify to pass after the update to capi PR

can you also add something to the docs about the bastion and how to use it?

**Comment by whites11** [comment]
> > @whites11 you are going to have to generate again to get the verify to pass after the update to capi PR
> 
> can you also add something to the docs about the bastion and how to use it?

Sure, mind suggesting where I should put those docs?

**Comment by nader-ziada** [comment]
> > > @whites11 you are going to have to generate again to get the verify to pass after the update to capi PR
> > 
> > 
> > can you also add something to the docs about the bastion and how to use it?
> 
> Sure, mind suggesting where I should put those docs?

you can either add a section in the private cluster section https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/book/src/topics/api-server-endpoint.md , or add a new topic https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/master/docs/book/src/topics

**Comment by whites11** [comment]
> > > > @whites11 you are going to have to generate again to get the verify to pass after the update to capi PR
> > > 
> > > 
> > > can you also add something to the docs about the bastion and how to use it?
> > 
> > 
> > Sure, mind suggesting where I should put those docs?
> 
> you can either add a section in the private cluster section https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/book/src/topics/api-server-endpoint.md , or add a new topic https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/master/docs/book/src/topics

ok Added a new topic here https://github.com/whites11/cluster-api-provider-azure/blob/azure-bastion/docs/book/src/topics/ssh-access.md

**Comment by nader-ziada** [comment]
thanks for adding the docs @whites11 

/lgtm
/approve

**Pull request merged**
